### PR TITLE
Multiple features

### PR DESCRIPTION
**Pull request by delthas**
_Several commits in one PR for easier rebasing. See individual commits for details._

_Not that Github may show some commits in the wrong order (ordered by date rather than by parent)._

### CODE REVIEW

File: service.go [hunk 1]
```diff
@@ -199,7 +207,16 @@ func handleServiceNetworkStatus(dc *downstreamConn, params []string) error {
...
 			details = fmt.Sprintf("%v channels", len(uc.channels))
 		} else {
-			statuses = append(statuses, "disconnected")
+			var lastError error
+			net.lock.Lock()
+			lastError = net.lastError
+			net.lock.Unlock()
+
+			if lastError != nil {
+				statuses = append(statuses, fmt.Sprintf("disconnected (%v)", lastError))
```
Comment (emersion) [inline service.go:222]: Maybe we can use `details` to not clutter the statuses wit the error message?

File: service.go [hunk 2]
```diff
@@ -199,7 +207,16 @@ func handleServiceNetworkStatus(dc *downstreamConn, params []string) error {
 			statuses = append(statuses, "connected as "+uc.nick)
 			details = fmt.Sprintf("%v channels", len(uc.channels))
 		} else {
-			statuses = append(statuses, "disconnected")
+			var lastError error
+			net.lock.Lock()
```
Comment (emersion) [inline service.go:217]: The service commands are handled from the user goroutine, so no lock should be required.

File: user.go [hunk 3]
```diff
@@ -43,8 +54,9 @@ type network struct {
 
 	history map[string]uint64
 
-	lock sync.Mutex
-	conn *upstreamConn
+	lock      sync.Mutex
+	conn      *upstreamConn
+	lastError error
```
Comment (emersion) [inline user.go:59]: (If this doesn't require a lock, let's move it above `lock`, alongside `history`)

File: user.go [hunk 4]
```diff
@@ -57,6 +69,15 @@ func newNetwork(user *user, record *Network) *network {
 	}
 }
 
+func (net *network) forEachDownstream(f func(*downstreamConn)) {
```
Comment (emersion) [inline user.go:72]: Maybe we can change `upstreamConn.forEachDownstream` to use this?

**Comment by delthas** [comment]
Updated.

**Pull request closed without merging**